src/storage/lsm.rs

```diff
@@ -200,18 +207,76 @@ impl LSM {
     }
 
     fn compact_sstables(&mut self) -> Result<()> {
-        let mut dense_idx = Memtable::default();
-        for old_sst in self.sstables.iter() {
-            dense_idx.update_from_commit_log(&old_sst.path)?;
+        let path = utils::timestamped_path(SSTABLES_DIR_PATH);
+        let mut file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open(&path)?;
+
+        let mut key_value_iterators = Vec::new();
+        for (index, table) in self.sstables.iter().enumerate() {
+            let p = &table.path;
+            let file = File::open(p)?;
+
+            // NB: the index/position of the sstable is included for the purpose of breaking ties
+            // on duplicate keys.
+            let iter = KeyValueIterator::from(file).zip(std::iter::repeat(index));
+            key_value_iterators.push(iter);
         }
-        let new_sst =
-            SSTable::write_from_memtable(&dense_idx, utils::timestamped_path(SSTABLES_DIR_PATH))?;
+
+        let compacted = key_value_iterators
+            .into_iter()
+            .kmerge_by(|(a, index_a), (b, index_b)| {
+                /*
+                the comparator contract dictates we return true iff |a| is ordered before |b|
+                    or said differently: |a| < |b|.
+
+                for equal keys, we define |a| < |b| iff |a| is more recent.
+                    note: |a| is more recent when index_a > index_b.
+
+                by defining the ordering in this way,
+                    we only keep the first instance of key |k| in the compacted iterator.
+                    duplicate items with key |k| must be discarded.
+                 */
+
+                // guide results to the front of the iterator for early termination
+                if a.is_err() {
+                    return true;
+                }
+                if b.is_err() {
+                    return false;
+                }
+
+                let key_a = &a.as_ref().unwrap().0;
+                let key_b = &b.as_ref().unwrap().0;
+
+                let a_is_equal_but_more_recent = key_a == key_b && index_a > index_b;
+                return key_a < key_b || a_is_equal_but_more_recent;
+            })
+            .map(|a| a.0); // tables[i] is no longer needed
+                           // .unique_by(|(k, _)| k.0.clone()); // keep first instance of |k|
```

Comment:
Why are we deduping manually? Something to do with borrow check? And if so, delete this line?

Reply:
Ah, yeah. It's because the items are results. This complicates the uniqueness check because we wish for the uniqueness to be determined by the string value of the key, but Errors don't have valid string value representations. It's an "result iterator" challenge which I haven't found a better solution for.

Reply:
I see